In [1]:
from backtest_bot import Backtest
import pandas as pd
from pathlib import Path
from datetime import datetime

##### Back end configurations for the **Backtest Bot**

- Set the data_folder
- Change the YAML configuration file to the file containing strategies you want to test. 

In [2]:
data_folder = Path('/Users/frederickmacdonald/Documents/Projects/true_botist/true-bautist/config')
yaml_config = 'extended_backtesting.yaml'
cfg = data_folder / yaml_config


##### Load the config file into the Backtest class and run the analysis

In [3]:
backtest = Backtest(cfg)
results = backtest.run()


In [6]:
# Get Performance Metrics for each strategy contained in the YAML Config file. 
backtest.get_performance_metrics()

{'ma_crossover_aapl': {'total_return': -6.766911297499112,
  'annual_return': -1.4275792862841752,
  'avg_daily_return': 0.01189067697501362,
  'return_std': 1.8764238752183189,
  'drawdown_metrics': {'max_drawdown': -0.5689134300993286,
   'avg_drawdown': -0.3461823211630413,
   'max_drawdown_duration': 1059,
   'avg_drawdown_duration': 486.9836347975883,
   'drawdown_frequency': 0.9454397394136808},
  'trade_metrics': {'total_trades': 26,
   'win_rate': 0.36,
   'avg_trade_return': 0.0024615418477049555,
   'best_trade': 0.4463746575388491,
   'worst_trade': -0.17583262964444624,
   'avg_winning_trade': 0.12344020834725436,
   'avg_losing_trade': -0.0655889580582916,
   'profit_factor': 1.0586403451267017},
  'risk_metrics': {'sharpe_ratio': 0.034112184379228666,
   'sortino_ratio': 0.048167403035790035,
   'calmar_ratio': 0.052669711052176625,
   'volatility': 0.297873055678306,
   'skewness': 0.06377360551401492,
   'kurtosis': 4.680358734952003,
   'var_95': -0.029085253312574433,

#### Results Breakdown

##### 1. For the *Bollinger Portfolio* (best performer):

**Return Metrics**:

- Total Return: +19.87% (total profit)
- Annual Return: +3.79% per year
- Average Daily Return: +0.017% per day
- Return Standard Deviation: 0.69% daily volatility


**Trade Performance**:


* Win Rate: 74.47% (percentage of profitable trades)
* Total Trades: 142 trades
* Best Trade: +11.98%
* Worst Trade: -5.41%
* Profit Factor: 3.79 (gross profits divided by gross losses - very good)


**Risk Metrics**:


* Sharpe Ratio: 0.21 (risk-adjusted returns - positive but low)
* Maximum Drawdown: -9.24% (biggest peak-to-trough decline)
* Value at Risk (95%): 0%
* Volatility: 10.89% annualized

##### 2. For the *RSI Tech Basket* (moderate performer):

**Return Metrics**:


- Total Return: +3.66%
- Annual Return: +0.74%
- Average Daily Return: +0.008%


**Trade Performance**:


* Win Rate: 69.52%
* Total Trades: 188 trades
* Best Trade: +9.31%
* Worst Trade: -7.56%
* Profit Factor: 2.42 (good)

##### 3. For the MA Crossover AAPL (worst performer):

**Return Metrics**:


* Total Return: -6.77% (loss)
* Annual Return: -1.43%
* Average Daily Return: +0.012%


**Trade Performance**:


- Win Rate: 36% (poor)
- Total Trades: 26 trades
- Best Trade: +44.64%
- Worst Trade: -17.58%
- Profit Factor: 1.06 (barely profitable)

The Bollinger Portfolio strategy performed best with:

         Highest total return
         Best win rate
         Best profit factor
         Smallest maximum drawdown
         Most consistent performance metrics

In [7]:
backtest._get_all_symbols()

['GOOGL', 'MSFT', 'AAPL']

In [10]:
backtest.strategies

{'ma_crossover_aapl': <Strategy.MovingAverageCrossover at 0x137b39b50>,
 'rsi_tech_basket': <Strategy.RSIStrategy at 0x1504ba4d0>,
 'bollinger_portfolio': <Strategy.BollingerBandsStrategy at 0x150b8f5d0>}

In [11]:
backtest.config

{'data': {'universe': {'type': 'static', 'symbols': ['AAPL', 'MSFT', 'GOOGL']},
  'parameters': {'interval': '1d',
   'history_start': '2020-01-01',
   'add_indicators': True,
   'cache_data': True,
   'adjustment': 'split+div'}},
 'strategies': {'ma_crossover_aapl': {'type': 'MovingAverageCrossover',
   'enabled': True,
   'symbols': ['AAPL'],
   'parameters': {'short_window': 20, 'long_window': 50}},
  'rsi_tech_basket': {'type': 'RSIStrategy',
   'enabled': True,
   'symbols': ['MSFT', 'GOOGL'],
   'parameters': {'period': 14, 'overbought': 70, 'oversold': 30}},
  'bollinger_portfolio': {'type': 'BollingerBandsStrategy',
   'enabled': True,
   'symbols': ['AAPL', 'MSFT', 'GOOGL'],
   'parameters': {'period': 20, 'std_dev': 2, 'mean_reversion': True}}},
 'portfolio': {'initial_capital': 100000,
  'position_sizing': {'method': 'equal',
   'max_position_size': 0.2,
   'max_correlation': 0.7},
  'rebalancing': {'frequency': 'monthly', 'threshold': 0.05}},
 'risk_management': {'max_portf